compiler/src/dotty/tools/dotc/typer/Inferencing.scala

```diff
@@ -167,33 +167,36 @@ object Inferencing {
 
     private var toMaximize: List[TypeVar] = Nil
 
-    def apply(x: Boolean, tp: Type): Boolean = tp.dealias match {
-      case _: WildcardType | _: ProtoType =>
-        false
-      case tvar: TypeVar if !tvar.isInstantiated =>
-        force.appliesTo(tvar)
-        && ctx.typerState.constraint.contains(tvar)
-        && {
-          val direction = instDirection(tvar.origin)
-          if minimizeSelected then
-            if direction <= 0 && tvar.hasNonWildcardLowerBound then
+    def apply(x: Boolean, tp: Type): Boolean =
+      try tp.dealias match
+        case _: WildcardType | _: ProtoType =>
+          false
+        case tvar: TypeVar if !tvar.isInstantiated =>
+          force.appliesTo(tvar)
+          && ctx.typerState.constraint.contains(tvar)
+          && {
+            val direction = instDirection(tvar.origin)
+            if minimizeSelected then
+              if direction <= 0 && tvar.hasNonWildcardLowerBound then
+                instantiate(tvar, fromBelow = true)
+              else if direction >= 0 && tvar.hasNonWildcardUpperBound then
+                instantiate(tvar, fromBelow = false)
+              // else hold off instantiating unbounded unconstrained variable
+            else if direction != 0 then
+              instantiate(tvar, fromBelow = direction < 0)
+            else if variance >= 0 && (force.ifBottom == IfBottom.ok || tvar.hasLowerBound) then
               instantiate(tvar, fromBelow = true)
-            else if direction >= 0 && tvar.hasNonWildcardUpperBound then
-              instantiate(tvar, fromBelow = false)
-            // else hold off instantiating unbounded unconstrained variable
-          else if direction != 0 then
-            instantiate(tvar, fromBelow = direction < 0)
-          else if variance >= 0 && (force.ifBottom == IfBottom.ok || tvar.hasLowerBound) then
-            instantiate(tvar, fromBelow = true)
-          else if variance >= 0 && force.ifBottom == IfBottom.fail then
-            return false
-          else
-            toMaximize = tvar :: toMaximize
-          foldOver(x, tvar)
-        }
-      case tp =>
-        foldOver(x, tp)
-    }
+            else if variance >= 0 && force.ifBottom == IfBottom.fail then
+              return false
+            else
+              toMaximize = tvar :: toMaximize
+            foldOver(x, tvar)
+          }
+        case tp =>
+          foldOver(x, tp)
+      catch case ex: Throwable =>
+        handleRecursive("traverse ", s"${tp.show}" , ex)
```

Comment:
Might be useful to print the constraints here too, since the loop might be due to cyclic constraints as we traverse type variables.

Reply:
Here you go. With that we get:
```
-- Error: i12640.scala:15:42 ---------------------------------------------------
15 |         case Some(s1,a) => Cons(a, () => unfold(s1,f))
   |                                          ^
   |Recursion limit exceeded.
   |Maybe there is an illegal cyclic reference?
   |If that's not the case, you could also try to increase the stacksize using the -Xss JVM option.
   |A recurring operation is (inner to outer):
   |
   |  traverse  (((S, S => F[Option[(S, T)]])) => 
   |  x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |    Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |  ]
   |) => R in Constraint(
   | uninstantiated variables: 
   | constrained types: [A, B](fa: F[A])(f: A => F[B]): F[B], 
   |  [F[_$3], T](head: T, tailFun: () => F[x.CpsStream[F, T]]): x.Cons[F, T]
   |, 
   |  [S, F[_$5], T]
   |    (s0: S)
   |      (f: S => F[Option[(S, T)]])
   |        (implicit evidence$1: x.CpsMonad[F]): F[x.CpsStream[F, T]]
   | bounds: 
   |     A := Option[(S, T)]
   |     B := x.CpsStream[F, T]
   |     F := 
   |    [R] =>> (((S, S => F[Option[(S, T)]])) => 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |        Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |      ]
   |    ) => R
   |     T
   |     >: (a : T) | 
   |      x.CpsStream[
   |        [T] =>> 
   |          x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R)
   |            , 
   |          T]
   |      , Nothing]
   |     S := (S, S => F[Option[(S, T)]])
   |     F := 
   |    [T] =>> 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), T]
   |     T
   | ordering: 
   |)
   |  traverse  ([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R) in Constraint(
   | uninstantiated variables: 
   | constrained types: [A, B](fa: F[A])(f: A => F[B]): F[B], 
   |  [F[_$3], T](head: T, tailFun: () => F[x.CpsStream[F, T]]): x.Cons[F, T]
   |, 
   |  [S, F[_$5], T]
   |    (s0: S)
   |      (f: S => F[Option[(S, T)]])
   |        (implicit evidence$1: x.CpsMonad[F]): F[x.CpsStream[F, T]]
   | bounds: 
   |     A := Option[(S, T)]
   |     B := x.CpsStream[F, T]
   |     F := 
   |    [R] =>> (((S, S => F[Option[(S, T)]])) => 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |        Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |      ]
   |    ) => R
   |     T
   |     >: (a : T) | 
   |      x.CpsStream[
   |        [T] =>> 
   |          x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R)
   |            , 
   |          T]
   |      , Nothing]
   |     S := (S, S => F[Option[(S, T)]])
   |     F := 
   |    [T] =>> 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), T]
   |     T
   | ordering: 
   |)
   |  traverse  x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |  Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |] in Constraint(
   | uninstantiated variables: 
   | constrained types: [A, B](fa: F[A])(f: A => F[B]): F[B], 
   |  [F[_$3], T](head: T, tailFun: () => F[x.CpsStream[F, T]]): x.Cons[F, T]
   |, 
   |  [S, F[_$5], T]
   |    (s0: S)
   |      (f: S => F[Option[(S, T)]])
   |        (implicit evidence$1: x.CpsMonad[F]): F[x.CpsStream[F, T]]
   | bounds: 
   |     A := Option[(S, T)]
   |     B := x.CpsStream[F, T]
   |     F := 
   |    [R] =>> (((S, S => F[Option[(S, T)]])) => 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |        Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |      ]
   |    ) => R
   |     T
   |     >: (a : T) | 
   |      x.CpsStream[
   |        [T] =>> 
   |          x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R)
   |            , 
   |          T]
   |      , Nothing]
   |     S := (S, S => F[Option[(S, T)]])
   |     F := 
   |    [T] =>> 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), T]
   |     T
   | ordering: 
   |)
   |  traverse  ((S, S => F[Option[(S, T)]])) => 
   |  x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |    Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |  ] in Constraint(
   | uninstantiated variables: 
   | constrained types: [A, B](fa: F[A])(f: A => F[B]): F[B], 
   |  [F[_$3], T](head: T, tailFun: () => F[x.CpsStream[F, T]]): x.Cons[F, T]
   |, 
   |  [S, F[_$5], T]
   |    (s0: S)
   |      (f: S => F[Option[(S, T)]])
   |        (implicit evidence$1: x.CpsMonad[F]): F[x.CpsStream[F, T]]
   | bounds: 
   |     A := Option[(S, T)]
   |     B := x.CpsStream[F, T]
   |     F := 
   |    [R] =>> (((S, S => F[Option[(S, T)]])) => 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |        Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |      ]
   |    ) => R
   |     T
   |     >: (a : T) | 
   |      x.CpsStream[
   |        [T] =>> 
   |          x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R)
   |            , 
   |          T]
   |      , Nothing]
   |     S := (S, S => F[Option[(S, T)]])
   |     F := 
   |    [T] =>> 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), T]
   |     T
   | ordering: 
   |)
   |  traverse  (((S, S => F[Option[(S, T)]])) => 
   |  x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |    Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |  ]
   |) => R in Constraint(
   | uninstantiated variables: 
   | constrained types: [A, B](fa: F[A])(f: A => F[B]): F[B], 
   |  [F[_$3], T](head: T, tailFun: () => F[x.CpsStream[F, T]]): x.Cons[F, T]
   |, 
   |  [S, F[_$5], T]
   |    (s0: S)
   |      (f: S => F[Option[(S, T)]])
   |        (implicit evidence$1: x.CpsMonad[F]): F[x.CpsStream[F, T]]
   | bounds: 
   |     A := Option[(S, T)]
   |     B := x.CpsStream[F, T]
   |     F := 
   |    [R] =>> (((S, S => F[Option[(S, T)]])) => 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |        Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |      ]
   |    ) => R
   |     T
   |     >: (a : T) | 
   |      x.CpsStream[
   |        [T] =>> 
   |          x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R)
   |            , 
   |          T]
   |      , Nothing]
   |     S := (S, S => F[Option[(S, T)]])
   |     F := 
   |    [T] =>> 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), T]
   |     T
   | ordering: 
   |)
   |  traverse  ([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R) in Constraint(
   | uninstantiated variables: 
   | constrained types: [A, B](fa: F[A])(f: A => F[B]): F[B], 
   |  [F[_$3], T](head: T, tailFun: () => F[x.CpsStream[F, T]]): x.Cons[F, T]
   |, 
   |  [S, F[_$5], T]
   |    (s0: S)
   |      (f: S => F[Option[(S, T)]])
   |        (implicit evidence$1: x.CpsMonad[F]): F[x.CpsStream[F, T]]
   | bounds: 
   |     A := Option[(S, T)]
   |     B := x.CpsStream[F, T]
   |     F := 
   |    [R] =>> (((S, S => F[Option[(S, T)]])) => 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |        Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |      ]
   |    ) => R
   |     T
   |     >: (a : T) | 
   |      x.CpsStream[
   |        [T] =>> 
   |          x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R)
   |            , 
   |          T]
   |      , Nothing]
   |     S := (S, S => F[Option[(S, T)]])
   |     F := 
   |    [T] =>> 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), T]
   |     T
   | ordering: 
   |)
   |  traverse  x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |  Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |] in Constraint(
   | uninstantiated variables: 
   | constrained types: [A, B](fa: F[A])(f: A => F[B]): F[B], 
   |  [F[_$3], T](head: T, tailFun: () => F[x.CpsStream[F, T]]): x.Cons[F, T]
   |, 
   |  [S, F[_$5], T]
   |    (s0: S)
   |      (f: S => F[Option[(S, T)]])
   |        (implicit evidence$1: x.CpsMonad[F]): F[x.CpsStream[F, T]]
   | bounds: 
   |     A := Option[(S, T)]
   |     B := x.CpsStream[F, T]
   |     F := 
   |    [R] =>> (((S, S => F[Option[(S, T)]])) => 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |        Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |      ]
   |    ) => R
   |     T
   |     >: (a : T) | 
   |      x.CpsStream[
   |        [T] =>> 
   |          x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R)
   |            , 
   |          T]
   |      , Nothing]
   |     S := (S, S => F[Option[(S, T)]])
   |     F := 
   |    [T] =>> 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), T]
   |     T
   | ordering: 
   |)
   |  traverse  ((S, S => F[Option[(S, T)]])) => 
   |  x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |    Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |  ] in Constraint(
   | uninstantiated variables: 
   | constrained types: [A, B](fa: F[A])(f: A => F[B]): F[B], 
   |  [F[_$3], T](head: T, tailFun: () => F[x.CpsStream[F, T]]): x.Cons[F, T]
   |, 
   |  [S, F[_$5], T]
   |    (s0: S)
   |      (f: S => F[Option[(S, T)]])
   |        (implicit evidence$1: x.CpsMonad[F]): F[x.CpsStream[F, T]]
   | bounds: 
   |     A := Option[(S, T)]
   |     B := x.CpsStream[F, T]
   |     F := 
   |    [R] =>> (((S, S => F[Option[(S, T)]])) => 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |        Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |      ]
   |    ) => R
   |     T
   |     >: (a : T) | 
   |      x.CpsStream[
   |        [T] =>> 
   |          x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R)
   |            , 
   |          T]
   |      , Nothing]
   |     S := (S, S => F[Option[(S, T)]])
   |     F := 
   |    [T] =>> 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), T]
   |     T
   | ordering: 
   |)
   |  traverse  (((S, S => F[Option[(S, T)]])) => 
   |  x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |    Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |  ]
   |) => R in Constraint(
   | uninstantiated variables: 
   | constrained types: [A, B](fa: F[A])(f: A => F[B]): F[B], 
   |  [F[_$3], T](head: T, tailFun: () => F[x.CpsStream[F, T]]): x.Cons[F, T]
   |, 
   |  [S, F[_$5], T]
   |    (s0: S)
   |      (f: S => F[Option[(S, T)]])
   |        (implicit evidence$1: x.CpsMonad[F]): F[x.CpsStream[F, T]]
   | bounds: 
   |     A := Option[(S, T)]
   |     B := x.CpsStream[F, T]
   |     F := 
   |    [R] =>> (((S, S => F[Option[(S, T)]])) => 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |        Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |      ]
   |    ) => R
   |     T
   |     >: (a : T) | 
   |      x.CpsStream[
   |        [T] =>> 
   |          x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R)
   |            , 
   |          T]
   |      , Nothing]
   |     S := (S, S => F[Option[(S, T)]])
   |     F := 
   |    [T] =>> 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), T]
   |     T
   | ordering: 
   |)
   |  traverse  ([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R) in Constraint(
   | uninstantiated variables: 
   | constrained types: [A, B](fa: F[A])(f: A => F[B]): F[B], 
   |  [F[_$3], T](head: T, tailFun: () => F[x.CpsStream[F, T]]): x.Cons[F, T]
   |, 
   |  [S, F[_$5], T]
   |    (s0: S)
   |      (f: S => F[Option[(S, T)]])
   |        (implicit evidence$1: x.CpsMonad[F]): F[x.CpsStream[F, T]]
   | bounds: 
   |     A := Option[(S, T)]
   |     B := x.CpsStream[F, T]
   |     F := 
   |    [R] =>> (((S, S => F[Option[(S, T)]])) => 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |        Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |      ]
   |    ) => R
   |     T
   |     >: (a : T) | 
   |      x.CpsStream[
   |        [T] =>> 
   |          x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R)
   |            , 
   |          T]
   |      , Nothing]
   |     S := (S, S => F[Option[(S, T)]])
   |     F := 
   |    [T] =>> 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), T]
   |     T
   | ordering: 
   |)
   |  ...
   |
   |  traverse  x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |  Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |] in Constraint(
   | uninstantiated variables: 
   | constrained types: [A, B](fa: F[A])(f: A => F[B]): F[B], 
   |  [F[_$3], T](head: T, tailFun: () => F[x.CpsStream[F, T]]): x.Cons[F, T]
   |, 
   |  [S, F[_$5], T]
   |    (s0: S)
   |      (f: S => F[Option[(S, T)]])
   |        (implicit evidence$1: x.CpsMonad[F]): F[x.CpsStream[F, T]]
   | bounds: 
   |     A := Option[(S, T)]
   |     B := x.CpsStream[F, T]
   |     F := 
   |    [R] =>> (((S, S => F[Option[(S, T)]])) => 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |        Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |      ]
   |    ) => R
   |     T
   |     >: (a : T) | 
   |      x.CpsStream[
   |        [T] =>> 
   |          x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R)
   |            , 
   |          T]
   |      , Nothing]
   |     S := (S, S => F[Option[(S, T)]])
   |     F := 
   |    [T] =>> 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), T]
   |     T
   | ordering: 
   |)
   |  traverse  ((S, S => F[Option[(S, T)]])) => 
   |  x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |    Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |  ] in Constraint(
   | uninstantiated variables: 
   | constrained types: [A, B](fa: F[A])(f: A => F[B]): F[B], 
   |  [F[_$3], T](head: T, tailFun: () => F[x.CpsStream[F, T]]): x.Cons[F, T]
   |, 
   |  [S, F[_$5], T]
   |    (s0: S)
   |      (f: S => F[Option[(S, T)]])
   |        (implicit evidence$1: x.CpsMonad[F]): F[x.CpsStream[F, T]]
   | bounds: 
   |     A := Option[(S, T)]
   |     B := x.CpsStream[F, T]
   |     F := 
   |    [R] =>> (((S, S => F[Option[(S, T)]])) => 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |        Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |      ]
   |    ) => R
   |     T
   |     >: (a : T) | 
   |      x.CpsStream[
   |        [T] =>> 
   |          x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R)
   |            , 
   |          T]
   |      , Nothing]
   |     S := (S, S => F[Option[(S, T)]])
   |     F := 
   |    [T] =>> 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), T]
   |     T
   | ordering: 
   |)
   |  traverse  (((S, S => F[Option[(S, T)]])) => 
   |  x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |    Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |  ]
   |) => R in Constraint(
   | uninstantiated variables: 
   | constrained types: [A, B](fa: F[A])(f: A => F[B]): F[B], 
   |  [F[_$3], T](head: T, tailFun: () => F[x.CpsStream[F, T]]): x.Cons[F, T]
   |, 
   |  [S, F[_$5], T]
   |    (s0: S)
   |      (f: S => F[Option[(S, T)]])
   |        (implicit evidence$1: x.CpsMonad[F]): F[x.CpsStream[F, T]]
   | bounds: 
   |     A := Option[(S, T)]
   |     B := x.CpsStream[F, T]
   |     F := 
   |    [R] =>> (((S, S => F[Option[(S, T)]])) => 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |        Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |      ]
   |    ) => R
   |     T
   |     >: (a : T) | 
   |      x.CpsStream[
   |        [T] =>> 
   |          x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R)
   |            , 
   |          T]
   |      , Nothing]
   |     S := (S, S => F[Option[(S, T)]])
   |     F := 
   |    [T] =>> 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), T]
   |     T
   | ordering: 
   |)
   |  traverse  [R] =>> (((S, S => F[Option[(S, T)]])) => 
   |  x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |    Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |  ]
   |) => R in Constraint(
   | uninstantiated variables: 
   | constrained types: [A, B](fa: F[A])(f: A => F[B]): F[B], 
   |  [F[_$3], T](head: T, tailFun: () => F[x.CpsStream[F, T]]): x.Cons[F, T]
   |, 
   |  [S, F[_$5], T]
   |    (s0: S)
   |      (f: S => F[Option[(S, T)]])
   |        (implicit evidence$1: x.CpsMonad[F]): F[x.CpsStream[F, T]]
   | bounds: 
   |     A := Option[(S, T)]
   |     B := x.CpsStream[F, T]
   |     F := 
   |    [R] =>> (((S, S => F[Option[(S, T)]])) => 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |        Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |      ]
   |    ) => R
   |     T
   |     >: (a : T) | 
   |      x.CpsStream[
   |        [T] =>> 
   |          x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R)
   |            , 
   |          T]
   |      , Nothing]
   |     S := (S, S => F[Option[(S, T)]])
   |     F := 
   |    [T] =>> 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), T]
   |     T
   | ordering: 
   |)
   |  traverse  ([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R) in Constraint(
   | uninstantiated variables: 
   | constrained types: [A, B](fa: F[A])(f: A => F[B]): F[B], 
   |  [F[_$3], T](head: T, tailFun: () => F[x.CpsStream[F, T]]): x.Cons[F, T]
   |, 
   |  [S, F[_$5], T]
   |    (s0: S)
   |      (f: S => F[Option[(S, T)]])
   |        (implicit evidence$1: x.CpsMonad[F]): F[x.CpsStream[F, T]]
   | bounds: 
   |     A := Option[(S, T)]
   |     B := x.CpsStream[F, T]
   |     F := 
   |    [R] =>> (((S, S => F[Option[(S, T)]])) => 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |        Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |      ]
   |    ) => R
   |     T
   |     >: (a : T) | 
   |      x.CpsStream[
   |        [T] =>> 
   |          x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R)
   |            , 
   |          T]
   |      , Nothing]
   |     S := (S, S => F[Option[(S, T)]])
   |     F := 
   |    [T] =>> 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), T]
   |     T
   | ordering: 
   |)
   |  traverse  x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), T] in Constraint(
   | uninstantiated variables: 
   | constrained types: [A, B](fa: F[A])(f: A => F[B]): F[B], 
   |  [F[_$3], T](head: T, tailFun: () => F[x.CpsStream[F, T]]): x.Cons[F, T]
   |, 
   |  [S, F[_$5], T]
   |    (s0: S)
   |      (f: S => F[Option[(S, T)]])
   |        (implicit evidence$1: x.CpsMonad[F]): F[x.CpsStream[F, T]]
   | bounds: 
   |     A := Option[(S, T)]
   |     B := x.CpsStream[F, T]
   |     F := 
   |    [R] =>> (((S, S => F[Option[(S, T)]])) => 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |        Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |      ]
   |    ) => R
   |     T
   |     >: (a : T) | 
   |      x.CpsStream[
   |        [T] =>> 
   |          x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R)
   |            , 
   |          T]
   |      , Nothing]
   |     S := (S, S => F[Option[(S, T)]])
   |     F := 
   |    [T] =>> 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), T]
   |     T
   | ordering: 
   |)
   |  traverse  [T] =>> 
   |  x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), T] in Constraint(
   | uninstantiated variables: 
   | constrained types: [A, B](fa: F[A])(f: A => F[B]): F[B], 
   |  [F[_$3], T](head: T, tailFun: () => F[x.CpsStream[F, T]]): x.Cons[F, T]
   |, 
   |  [S, F[_$5], T]
   |    (s0: S)
   |      (f: S => F[Option[(S, T)]])
   |        (implicit evidence$1: x.CpsMonad[F]): F[x.CpsStream[F, T]]
   | bounds: 
   |     A := Option[(S, T)]
   |     B := x.CpsStream[F, T]
   |     F := 
   |    [R] =>> (((S, S => F[Option[(S, T)]])) => 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |        Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |      ]
   |    ) => R
   |     T
   |     >: (a : T) | 
   |      x.CpsStream[
   |        [T] =>> 
   |          x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R)
   |            , 
   |          T]
   |      , Nothing]
   |     S := (S, S => F[Option[(S, T)]])
   |     F := 
   |    [T] =>> 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), T]
   |     T
   | ordering: 
   |)
   |  traverse  Nothing in Constraint(
   | uninstantiated variables: 
   | constrained types: [A, B](fa: F[A])(f: A => F[B]): F[B], 
   |  [F[_$3], T](head: T, tailFun: () => F[x.CpsStream[F, T]]): x.Cons[F, T]
   |, 
   |  [S, F[_$5], T]
   |    (s0: S)
   |      (f: S => F[Option[(S, T)]])
   |        (implicit evidence$1: x.CpsMonad[F]): F[x.CpsStream[F, T]]
   | bounds: 
   |     A := Option[(S, T)]
   |     B := x.CpsStream[F, T]
   |     F := 
   |    [R] =>> (((S, S => F[Option[(S, T)]])) => 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |        Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |      ]
   |    ) => R
   |     T
   |     >: (a : T) | 
   |      x.CpsStream[
   |        [T] =>> 
   |          x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R)
   |            , 
   |          T]
   |      , Nothing]
   |     S := (S, S => F[Option[(S, T)]])
   |     F := 
   |    [T] =>> 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), T]
   |     T
   | ordering: 
   |)
   |  traverse  Nothing[Option[((S, S => F[Option[(S, T)]]), Nothing)]] in Constraint(
   | uninstantiated variables: 
   | constrained types: [A, B](fa: F[A])(f: A => F[B]): F[B], 
   |  [F[_$3], T](head: T, tailFun: () => F[x.CpsStream[F, T]]): x.Cons[F, T]
   |, 
   |  [S, F[_$5], T]
   |    (s0: S)
   |      (f: S => F[Option[(S, T)]])
   |        (implicit evidence$1: x.CpsMonad[F]): F[x.CpsStream[F, T]]
   | bounds: 
   |     A := Option[(S, T)]
   |     B := x.CpsStream[F, T]
   |     F := 
   |    [R] =>> (((S, S => F[Option[(S, T)]])) => 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |        Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |      ]
   |    ) => R
   |     T
   |     >: (a : T) | 
   |      x.CpsStream[
   |        [T] =>> 
   |          x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R)
   |            , 
   |          T]
   |      , Nothing]
   |     S := (S, S => F[Option[(S, T)]])
   |     F := 
   |    [T] =>> 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), T]
   |     T
   | ordering: 
   |)
   |  traverse  ((S, S => F[Option[(S, T)]])) => 
   |  Nothing[Option[((S, S => F[Option[(S, T)]]), Nothing)]] in Constraint(
   | uninstantiated variables: 
   | constrained types: [A, B](fa: F[A])(f: A => F[B]): F[B], 
   |  [F[_$3], T](head: T, tailFun: () => F[x.CpsStream[F, T]]): x.Cons[F, T]
   |, 
   |  [S, F[_$5], T]
   |    (s0: S)
   |      (f: S => F[Option[(S, T)]])
   |        (implicit evidence$1: x.CpsMonad[F]): F[x.CpsStream[F, T]]
   | bounds: 
   |     A := Option[(S, T)]
   |     B := x.CpsStream[F, T]
   |     F := 
   |    [R] =>> (((S, S => F[Option[(S, T)]])) => 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), 
   |        Option[((S, S => F[Option[(S, T)]]), Nothing)]
   |      ]
   |    ) => R
   |     T
   |     >: (a : T) | 
   |      x.CpsStream[
   |        [T] =>> 
   |          x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R)
   |            , 
   |          T]
   |      , Nothing]
   |     S := (S, S => F[Option[(S, T)]])
   |     F := 
   |    [T] =>> 
   |      x.CpsStream[([R] =>> (((S, S => F[Option[(S, T)]])) => Nothing) => R), T]
   |     T
   | ordering: 
   |)
```
@rssh My gut feeling is that you are trying to push higher-kinded types too far here. I mean nobody stops you, but if you want code to be widely used then crazy types and constraints like this are a show-stopper.

Reply:
Interestingly, that original (unminimized) bug source is an attempt to create adapter for a well-known library [fs2 if anybody care],  so the genie was already released from the bottle a long time ago.